src/ssa/goto_unwinder.cpp

```diff
@@ -44,6 +46,121 @@ void goto_local_unwindert::reconnect_loops()
   }
 }
 
+std::string get_object_name(const exprt &expr)
+{
+  std::string name;
+  const auto &obj_op=to_address_of_expr(expr).op();
+  if(obj_op.id()==ID_symbol)
+    name=to_symbol_expr(obj_op).get_identifier().c_str();
+  else
+    name=to_symbol_expr(
+      to_index_expr(obj_op).array()).get_identifier().c_str();
+  return name;
+}
+
+/// Recursively renames dynamic objects in the expression tree.
+///
+/// \param it The instruction we are currently processing
+/// \param expr Subexpression to process.
+/// \param rename_map Objects which have been renamed so far.
+/// \pre Symbols of the previously defined dynamic objects have been cleared.
+void goto_local_unwindert::rename_dynamic_objects_rec(
+  goto_programt::instructiont &it,
+  exprt &expr,
+  std::map<std::string, std::string> &rename_map)
+{
+  if(expr.id()==ID_address_of)
+  {
+    std::string name=get_object_name(expr);
+    std::string suffix="$" + std::to_string(it.location_number);
+    // The form is dynamic_object$loc$<other suffixes>, keep the end suffix
+    auto pos=name.find('$');
+    if(pos!=std::string::npos)
+    {
+      auto second_pos=name.find('$', pos+1);
+      if(second_pos!=std::string::npos)
+        suffix+=name.substr(second_pos);
+    }
+    expr=create_dynamic_object(
+      suffix,
+      expr.type().subtype(),
+      goto_model.symbol_table,
+      suffix.find("$co")!=std::string::npos);
+    rename_map[name]=get_object_name(expr);
+  }
+  else if(expr.id()==ID_symbol)
+  {
+    // Rename already renamed dynamic objects in expressions
+    // This occurs in memsafety (--pointer-check) due to concrete objects
+    auto &symbol_expr=to_symbol_expr(expr);
+    auto rename=rename_map.find(symbol_expr.get_identifier().c_str());
+    if(rename!=rename_map.end())
+      symbol_expr.set_identifier(rename->second);
+  }
+  else if (expr.id()==ID_if && expr.get_bool("#malloc_result") &&
+           to_if_expr(expr).cond().id()==ID_and)
+  {
+    // Update the condition for concrete object selection based on new objects
+    // Its form is <nondet> && pointer_equalities, update only pointer eqs.
+    auto &if_expr=to_if_expr(expr);
+    // Get the object, its type and calculate new pointer equalities,
+    // we do not want the new objects to be included.
+    auto &obj=to_typecast_expr(if_expr.true_case()).op();
+    auto pointers=collect_pointer_vars(
+      goto_model.symbol_table,
+      obj.type().subtype());
```

Comment:
I'm not really sure. Since `collect_pointer_vars` is called before descending into true and false cases, the objects created in them are not in the symbolic table at this moment. Which means that fields of those objects do not make it to the new condition.

Consider this situation:
```
... = (do$0.next != &do$0$co) ? &do$0$co : &do$0
```
Let's say that `do$0` should be renamed to `do$1`. At the point of processing the condition, `&do$0` is not properly renamed, hence `do$1.next` won't appear in `pointers`. 

I may be missing something, but shouldn't we first descend into the conditions and only then call `collect_pointer_vars`?

Reply:
Is this a situation that can happen? Looking at https://github.com/diffblue/2ls/blob/master/src/ssa/malloc_ssa.cpp#L207 , the pointers are collected before objects are created for the given location. It probably won't hurt to try first recursing into true/false cases and then modifying the condition. It could possibly be the source of some of the false results, though I don't think that's the case (but memsafety runs quite quickly, so I will try and let you know).

Reply:
Interesting, without giving it much deep thought, I'd say that it may happen in every linked list (`next` of the object allocated in some previous iteration points to the concrete object, hence we cannot re-allocate it). In such case, the above would be a bug, although that should make most of the memsafety examples fail, so I must be missing something here.

Reply:
I tried it the other way around (first replace true/false and then collect pointer vars) and it did not influence anything, the score is the same. I don't know why but it seems that it does not really matter

Reply:
Ok, let's keep it, but it might be worth some investigation.